### DISCUSSION TITLE
Texture2D Submission Throwing 404?

### DISCUSSION THREAD

**Discussion by taylorbrown2686** [RageAgainstThePixel/com.openai.unity#259] [Q&A] [answered]
When I submit an image-based request:
```csharp
public async Task<string> GetImageBasedReply(string prompt, Texture2D image)
        {
            ChatSession.chatMessages.Add(new Message(Role.System, new List<Content> {
                new Content(ContentType.Text, "Use this image for context to the prompt."),
                image
            }));
            ChatSession.chatMessages.Add(new Message(Role.User, prompt));

            ChatRequest request = new ChatRequest(ChatSession.chatMessages, model: "gpt-4-vision-preview", maxTokens: 200);
            ChatResponse response = await ChatSession.client.ChatEndpoint.GetCompletionAsync(request);

            if (response.Choices is not null && response.Choices.Count > 0)
            {
                var gptMessage = response.Choices[0].Message;

                print("GPT Response: " + gptMessage.Content.ToString());

                ChatSession.chatMessages.Add(gptMessage);

                return gptMessage.Content.ToString();
            }
            else
            {
                return null;
            }
        }
```
        
It throws a [POST:404] RestException, simply saying it 'Failed' with no additional information. The error itself shows the preprompt given along with a long string (Base64 assumed?) that is truncated for length.

What exactly is the '404' referring to, and how can I resolve it?

**Comment by StephenHodgson** [comment]
`gpt-4-vision-preview` is no longer a valid model. Try using `gpt-4o`

Also if you could provide the full response output that would also be helpful.

**Reply by taylorbrown2686** [reply]
I did notice that after submission and changed over to gpt-4o. I also believe the account we had the API key tied to was expired in terms of payment, but that should be resolved now. We have gone from a 404 to a 400, so clearly something changed.

Here is the full output minus the B64 string:
```
RestException: [POST:400] https://api.openai.com/v1/chat/completions Failed!
{
  "request": {
    "url": "https://api.openai.com/v1/chat/completions",
    "body": {
      "messages": [
        {
          "role": "system",
          "content": "You are assisting a blind individual."
        },
        {
          "role": "system",
          "content": [
            {
              "type": "text",
              "text": "Use this image for context to the prompt."
            },
            {
              "type": "image_url",
              "image_url": {
                "url":
```

**Reply by StephenHodgson** [reply]
if 400 it should also include a reason why it failed.

**Reply by taylorbrown2686** [reply]
Hello, sorry for the late reply. I was busy with other projects. The 400 still occurs, here is the dump:

`RestException: [POST:400] https://api.openai.com/v1/chat/completions Failed!
{
  "request": {
    "url": "https://api.openai.com/v1/chat/completions",
    "body": {
      "messages": [
        {
          "role": "system",
          "content": "You may be asked to describe a scene in a video game. If you are, please assume it is real life and don't mention that it is clearly computer-generated graphics."
        },
        {
          "role": "system",
          "content": [
            {
              "type": "text",
              "text": "Use this image for context to the prompt."
            },
            {
              "type": "image_url",
              "image_url": {
                "url":`


I do not see any 'reason' provided, is there a way to get more verbose logs or is it dumped somewhere else possibly?

**Reply by StephenHodgson** [reply]
You can enable verbose debugging with `OpenAIClient.EnableDebug = true;`

**Reply by StephenHodgson** [reply]
400 typically means the request to the API is bad or malformed. Show me both request and response. Be sure to clean out any sensitive info in your logs before posting.

**Reply by taylorbrown2686** [reply]
Log output (EnableDebug: True)

```
RestException: [POST:400] https://api.openai.com/v1/chat/completions Failed!
{
  "request": {
    "url": "https://api.openai.com/v1/chat/completions",
    "body": {
      "messages": [
        {
          "role": "system",
          "content": "PROMPT HIDDEN"
        },
        {
          "role": "system",
          "content": [
            {
              "type": "text",
              "text": "PROMPT HIDDEN"
            },
            {
              "type": "image_url",
              "image_url": {
                "url":
```
Request/Response (called after image is captured, response returned from Task and processed elsewhere, but the response code here is where the failure is):
```csharp
public async Task<string> GetImageBasedReply(string prompt, Texture2D image)
        {
            ChatSession.chatMessages.Add(new Message(Role.System, new List<Content> {
                new Content(ContentType.Text, "PROMPT HIDDEN"),
                image
            }));
            ChatSession.chatMessages.Add(new Message(Role.User, prompt));

            ChatRequest request = new ChatRequest(ChatSession.chatMessages, OpenAI.Models.Model.GPT4o, maxTokens: 200);
            ChatResponse response = await ChatSession.client.ChatEndpoint.GetCompletionAsync(request);

            if (response.Choices is not null && response.Choices.Count > 0)
            {
                var gptMessage = response.Choices[0].Message;

                print("GPT Response: " + gptMessage.Content.ToString());

                ChatSession.chatMessages.Add(gptMessage);

                return gptMessage.Content.ToString();
            }
            else
            {
                return null;
            }
        }
```

**Reply by taylorbrown2686** [reply]
Sorry for the formatting issue, I am putting the whole string of code in the " ` " like it wants me to, but some of the code refuses to display properly.

**Reply by StephenHodgson** [reply]
it's ok I fixed it. you need 3x `

I don't see the full output of your log which is strange. Not sure why it is getting cut off. Try looking in editor log for the full string.

![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/13334553/69b62bdc-3e60-4c0e-aed2-595d37bdcc49)

**Reply by taylorbrown2686** [reply, marked as answer]
Hmm, I've never had to do that for a log message before. I didn't even know that option existed!

The B64 string was hiding the full message for the '400' error, which specifies that image queries need to be sent under the 'User' role and not the 'System' role. Changing this fixed my problem.

Thanks for the help!
